### PR TITLE
Periodically update DNS caches for better privacy of non-reachable nodes

### PR DESCRIPTION
**Pull request by naumenkogs**
Every time a fresh Bitcoin Core node starts, it makes a DNS query to learn about nodes in the network. This process leaks the privacy of those new nodes: every required DNS server and the corresponding infrastructure would be aware that a new node was spinned up in a particular internet segment.
The goal of this proposal is to reduce the number of those actors learning about a new node.
The way to achieve it is to keep DNS server caches updated, so that new nodes rarely hit anything past the early servers.

To keep them updated, every reachable node would periodically make all widely used DNS queries, thus, triggering DNS cache updates on the resolvers appearing on their path to the end DNS servers.
This, obviously, would leak more information about the existence of these reachable nodes. We think this is no big deal, because those nodes are already easy to find, since they are reachable.

Note that this helps only to those private nodes, which share a subnet with some reachable node running this code.

In future, it would be great to analyze the results of those queries against a local AddrMan to check for anomalies.

The idea was originally proposed by Greg Maxwell.

### CODE REVIEW

**Comment by sipa** [comment]
Can this use the scheduler instead of using a separate thread?

**Comment by practicalswift** [comment]
Concept ACK

**Comment by naumenkogs** [comment]
@sipa yes! I simply forgot that this can be done with scheduler. Just force-pushed an update.

File: doc/reduce-traffic.md [hunk 1]
```diff
@@ -28,7 +28,9 @@ calculating the target.
 
 Disabling listening will result in fewer nodes connected (remember the maximum of 8
 outbound peers). Fewer nodes will result in less traffic usage as you are relaying
-blocks and transactions to fewer nodes.
+blocks and transactions to fewer nodes. This would also prevent reachable nodes from
+periodically querying DNS servers when it is not needed (feature, which gives more
+privacy to the non-reachable nodes located in the same subnet).
```
Comment (MarcoFalke) [inline doc/reduce-traffic.md]: I think this in an implementation detail which people that want to reduce their traffic are not interested in learning about.

A release notes snippet would be a better place to put this, no?

Comment (naumenkogs) [reply doc/reduce-traffic.md]: Yeah, I was hesitant about this part. Committed this for now simply to not ignore this. Unless someone expresses support for keeping this, I would remove this commit later.

File: src/net.cpp [hunk 2]
```diff
@@ -1571,8 +1579,48 @@ void StopMapPort()
...
+    // Only reachable nodes do this, because those are already easy to find.
+    // Unless the node is explicitly set to be non-reachable (-listen=0),
+    // the only way to check if it is reachable is to observe inbound connections.
+    // This check is made every time, because a node may become non-reachable
+    // due to external factors (e.g. local network settings update).
+    if (CountInboundConnections() == 0) return;
+
+    // Every time a new widely-requests service bit is deployed, this list should be
+    // updated to invalide records for that service bit.
+    std::vector<uint64_t> service_bits_combinations{
```
Comment (MarcoFalke) [inline src/net.cpp]: nit: I think this list reads easier if each entry was on its own line.

nit: compile time constants should be all upper case and static

```suggestion
    static const std::vector<uint64_t> SERVICE_BITS_COMBINATIONS{
```

File: src/net.h [hunk 3]
```diff
@@ -468,6 +479,7 @@ class CConnman
 
     CThreadInterrupt interruptNet;
 
+    std::thread thread_dns_caches_update;
```
Comment (MarcoFalke) [inline src/net.h]: nit: Should be removed now?

File: src/net.h [hunk 4]
```diff
@@ -361,6 +362,16 @@ class CConnman
...
 
+    // Reachable nodes should periodically make requests
+    // to DNS servers even if they don't need them.
+    // Then, when new nodes from the same subnets need to query DNS,
+    // they are likely to hit the cache instead.
+    // For non-reachable nodes this means more privacy,
+    // because fewer actors would know about their query.
+    // For reachable nodes there is no privacy loss due to this,
+    // because being reachable already makes them public.
+    void DNSCachesUpdate();
```
Comment (MarcoFalke) [inline src/net.h]: nit: Both of the newly added methods can be `const` because they are read-only, no?
```suggestion
    void DNSCachesUpdate() const;
```

**Review by maflcko — COMMENTED** [review]
Just some style-nits

File: src/net.h (hunk 3 above)
Comment (naumenkogs) [reply src/net.h]: oops. yes!

**Comment by MarcoFalke** [comment]
> Can this use the scheduler instead of using a separate thread?

What is the maximum time this can take? If it is unbounded, it seems that this would be a trivial DOS attack against a node, no?

**Comment by naumenkogs** [comment]
@sipa I was not aware that scheduled event will block other execution threads, and now that Marco pointed out that attack, i'm thinking separate thread makes more sense? We don't want to make any assumptions about these requests executing very fast.

**Comment by MarcoFalke** [comment]
`LimitValidationInterfaceQueue` is called in `ActivateBestChain`. So even validation itself can be blocked, if an attacker can make those requests go slow. Not sure if that is possible, but I wanted to raise the point.

**Comment by sipa** [comment]
@naumenkogs It won't block other threads, but it will delay other events in the same scheduler. I agree that it's probably too slow for that (I don't know what latency `getaddrinfo` can have, but I expect it may be substantial). An alternative is doing it from `PeerValidationLogic::SendMessages`, once a timer expires. I'm hammering on avoiding a separate thread for this, because threads have a significant memory footprint (several MB, at least), so a thread that's idle almost all the time is a waste.

@MarcoFalke Vaguely related, do we have any policies around what maximum execution duration things in the main scheduler are allowed to have?

**Comment by naumenkogs** [comment]
Agreed that that RAM overhead of a new thread for just this is no good.

Yeah, I would say normal execution of this function takes 2s, abnormal may take...a minute even if everyone is honest, due to random internet/os glitches.
Do we have anything in our toolset to make sure Lookup() function doesn't take more than couple seconds? I don't see anything within scheduler. Maybe it's doable to add this feature?

An alternative crazy idea is to have a special thread for a number of non-so-critical functions, which might take long time (for example, I can imagine DumpAddresses takes long due to I/O issues).

File: src/net.cpp [hunk 5]
```diff
@@ -1571,8 +1579,53 @@ void StopMapPort()
...
 
 
+void CConnman::DNSCachesUpdate() const
+{
+    // Only reachable nodes do this, because those are already easy to find.
+    // Unless the node is explicitly set to be non-reachable (-listen=0),
+    // the only way to check if it is reachable is to observe inbound connections.
+    // This check is made every time, because a node may become non-reachable
+    // due to external factors (e.g. local network settings update).
+    if (CountInboundConnections() == 0) return;
```
Comment (luke-jr) [inline src/net.cpp:1589]: How about a `-dnscloaking` option to force it on/off?

File: src/net.cpp [hunk 6]
```diff
@@ -1571,8 +1579,53 @@ void StopMapPort()
...
+
+    // Every time a new widely-requests service bit is deployed, this list should be
+    // updated to invalide records for that service bit.
+    std::vector<uint64_t> SERVICE_BITS_COMBINATIONS{
+        // Legacy nodes
+        NODE_NETWORK,
```
Comment (luke-jr) [inline src/net.cpp:1595]: Wouldn't these be requesting the bare name without any service bit specification?

File: src/net.cpp [hunk 7]
```diff
@@ -1571,8 +1579,53 @@ void StopMapPort()
...
+        NODE_NETWORK_LIMITED,
```
Comment (luke-jr) [inline src/net.cpp:1596]: What nodes would ever request this?

File: src/net.cpp [hunk 8]
```diff
@@ -1571,8 +1579,53 @@ void StopMapPort()
...
+        }
 
+        // first query as if client doesn't support service bits first
+        std::vector<CNetAddr> ips;
+        LookupHost(seed, ips, 1, true);
+        // TODO: Here and below, make sure downloaded IPs are not very different
+        // from what stored locally, to identify AddrMan or DNS poisoning.
 
+        // query for all widely used combinations of service bits
+        for (uint64_t service_bits_combination : SERVICE_BITS_COMBINATIONS) {
```
Comment (luke-jr) [inline src/net.cpp:1619]: Should these really be done all at the same time? No new node would batch them like this...?

File: src/net.cpp [hunk 9]
```diff
@@ -2322,8 +2375,13 @@ bool CConnman::Start(CScheduler& scheduler, const Options& connOptions)
 
     if (!gArgs.GetBoolArg("-dnsseed", true))
         LogPrintf("DNS seeding disabled\n");
-    else
+    else {
         threadDNSAddressSeed = std::thread(&TraceThread<std::function<void()> >, "dnsseed", std::function<void()>(std::bind(&CConnman::ThreadDNSAddressSeed, this)));
+        if (fListen && fNameLookup && !HaveNameProxy()) {
+            LogPrintf("Proactive querying DNS servers to update caches is enabled\n");
+            scheduler.scheduleEvery([this] { DNSCachesUpdate(); }, DNS_CACHES_UPDATE_INTERVAL);
```
Comment (luke-jr) [inline src/net.cpp:2382]: Probably there should be some randomness added to the interval for each iteration?

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/net.cpp (hunk 5 above)
Comment (naumenkogs) [reply src/net.cpp:1589]: Good idea. Will add this config and add the documentation once we get more concept acks.

File: src/net.cpp (hunk 6 above)
Comment (naumenkogs) [reply src/net.cpp:1595]: I think this should be x1.dns.whatever.com?

File: src/net.cpp (hunk 8 above)
Comment (naumenkogs) [reply src/net.cpp:1619]: Well, one private node in the subnet will request one combination, another private node will request another one...

Why would you want the already public node separate cache updates in time?

**Comment by TheBlueMatt** [comment]
I thought we had two scheduler threads? If not we should just do that and then its not a problem.

File: src/net.cpp [hunk 10]
```diff
@@ -937,6 +940,16 @@ bool CConnman::AttemptToEvictConnection()
     return false;
 }
 
+int CConnman::CountInboundConnections() const
+{
+    int inbounds = 0;
+    LOCK(cs_vNodes);
+    for (const CNode* node : vNodes) {
```
Comment (TheBlueMatt) [inline src/net.cpp:947]: Can we count that such connections are from non-local addresses? Should be an extra two LoC

**Comment by naumenkogs** [comment]
Not sure what to do with this PR.
The show-stopper here was the discussion around the blocking behavior of DNS caches: the options were either add a new thread or employing existing threads for such non-critical task (see [discussion](https://github.com/bitcoin/bitcoin/issues/18488)).

It kind of went nowhere, so I'm gonna close this PR for now. If you have an opinion on the way to move forward with one of those paths, please go to the mentioned [issue](https://github.com/bitcoin/bitcoin/issues/18488).

If you have an alternative way to move forward, feel free to comment here and we can resurrect it.

**Pull request closed without merging**